include/zephyr/modem/chat.h

```diff
@@ -0,0 +1,350 @@
+	{                                                                                          \
+		.request = _request, .response_matches = _response_matches,                        \
+		.response_matches_size = ARRAY_SIZE(_response_matches), .timeout = 0,              \
+	}
+
+#define MODEM_CHAT_SCRIPT_CMD_RESP_NONE(_request, _timeout)                                        \
+	{                                                                                          \
+		.request = _request, .response_matches = NULL, .response_matches_size = 0,         \
+		.timeout = _timeout,                                                               \
+	}
+
+#define MODEM_CHAT_SCRIPT_CMDS_DEFINE(_sym, ...)                                                   \
+	const static struct modem_chat_script_chat _sym[] = {__VA_ARGS__}
+
+enum modem_chat_script_result {
+	MODEM_CHAT_SCRIPT_RESULT_SUCCESS,
+	MODEM_CHAT_SCRIPT_RESULT_ABORT,
+	MODEM_CHAT_SCRIPT_RESULT_TIMEOUT
+};
+
+/**
+ * @brief Callback called when script chat is received
+ *
+ * @param chat Pointer to chat instance instance
+ * @param result Result of script execution
+ * @param user_data Free to use user data set during modem_chat_init()
+ */
+typedef void (*modem_chat_script_callback)(struct modem_chat *chat,
+					   enum modem_chat_script_result result, void *user_data);
+
+/**
+ * @brief Modem chat script
+ *
+ * @param name Name of script
+ * @param script_chats Array of script chats
+ * @param script_chats_size Elements in array of script chats
+ * @param abort_matches Array of abort matches
+ * @param abort_matches_size Elements in array of abort matches
+ * @param callback Callback called when script execution terminates
+ * @param timeout Timeout in seconds within which the script execution must terminate
+ */
+struct modem_chat_script {
+	const char *name;
+	const struct modem_chat_script_chat *script_chats;
+	const uint16_t script_chats_size;
+	const struct modem_chat_match *const abort_matches;
+	const uint16_t abort_matches_size;
+	modem_chat_script_callback callback;
+	const uint32_t timeout;
+};
+
+#define MODEM_CHAT_SCRIPT_DEFINE(_sym, _script_chats, _abort_matches, _callback, _timeout)         \
+	static struct modem_chat_script _sym = {                                                   \
+		.name = #_sym,                                                                     \
+		.script_chats = _script_chats,                                                     \
+		.script_chats_size = ARRAY_SIZE(_script_chats),                                    \
+		.abort_matches = _abort_matches,                                                   \
+		.abort_matches_size = ARRAY_SIZE(_abort_matches),                                  \
+		.callback = _callback,                                                             \
+		.timeout = _timeout,                                                               \
+	}
```

Comment:
As far as I can see, one cannot change the request string of a `modem_chat_script_chat` during runtime. however, it might be necessary, when entering a command with mutable arguments (e.g. entering the sim pin). 
right now i would prepare the script on compile time and would `snprintf` the missing informations during runtine.

Reply:
yes, but the buffer the request member is pointing to is of fixed size and has to be allocated static (not on the stack) due to the constraints from the macros. Within the modem context implementation I could do stuff like:

```c
// example command for snprintf stuff
#define CMD_FMT "AT+SEND %s"
#define CMD_SZ sizeof("AT+SEND xxxx,xx,xxxx")
void send_msg(char* msg)
{
    char buf[CMD_SZ];
    size_t len = snprintf(buf, CMD_SZ, CMD_FMT, msg);
    int ret = modem_cmd_send(iface, cmd_handler, NULL, 0, buf, &sem_response, K_SECONDS(42));
    return ret;
}
```

This is extensively used in the hl7800 and wncm14a2a drivers. `modem_cmd_send` blocks until the cmd is send, so we can savely return. Obviously we cannot do that due to nature of the implementation using the kernel workq instead of a own rx thread and response semaphore.

Reply:
It is a deliberate design decision to not allow for ad-hock synchronization of commands, and to enforce the lifetime of the buffers passed to the chat scripts (forever), do you disagree with this design decision? Its hard to determine from your comment :)

Reply:
In any case, data sent to the modem is sent through a modem_pipe struct, which takes and returns raw bytes, it may be valuable to create another module to use in cases like this, where you may want to change the "AT mode" to data mode, which is used for stuff like socket offloading. Switching between the chat module and some raw data module (like PPP for example) is a supported usecase :)

Reply:
> do you disagree with this design decision? Its hard to determine from your comment :)
It was a value free observation. I like to keep the data/bss segments small, hence I like to use the stack. But - as you said - I had socket offloading on my mind... for this a raw data module might be as good as the chat module.